component/src/governance/check.rs

```diff
@@ -226,6 +305,91 @@ pub mod stateful {
         Ok(())
     }
 
+    async fn proposal_started_at_block_height<S: StateRead>(
+        state: S,
+        proposal_id: u64,
+        claimed_start_height: u64,
+    ) -> Result<()> {
+        if let Some(start_height) = state.proposal_voting_start(proposal_id).await? {
+            if start_height != claimed_start_height {
+                anyhow::bail!(
+                    "proposal {} was not started at claimed block height of {}",
+                    proposal_id,
+                    claimed_start_height
+                );
+            }
+        } else {
+            anyhow::bail!("proposal {} does not exist", proposal_id);
+        }
+
+        Ok(())
+    }
+
+    async fn nullifier_unspent_before<S: StateRead>(
+        state: S,
+        start_height: u64,
+        nullifier: &Nullifier,
+    ) -> Result<()> {
+        if let Some(spend_info) = state.spend_info(*nullifier).await? {
+            if spend_info.spend_height < start_height {
+                anyhow::bail!(
+                    "nullifier {} was already spent at block height {} before proposal started at block height {}",
+                    nullifier,
+                    spend_info.spend_height,
+                    start_height
+                );
+            }
+        }
+
+        Ok(())
+    }
```

Comment:
Would it make sense to commonize this with the shielded pool's `StateReadExt::check_nullifier_unspent`, if not exactly the same method (passing an `Option` to signal whether a start height is relevant seems ugly), at least in the same trait?